### PR TITLE
CoreCLR runtime tests + Mono LLVM AOT on arm64 Linux

### PR DESCRIPTION
**Pull request by imhameed**
This change:

- Adds new options to mono.proj:

    - MonoAOTEnableLLVM, which enables (or disables) building a Mono AOT
    cross compiler with LLVM;

    - MonoAOTLLVMDir, which specifies the path to a copy of LLVM suitable for
    building a Mono AOT cross compiler, and is optional; and

    - BuildMonoAOTCrossCompilerOnly, which allows building a Mono AOT
    cross compiler without building an associated full Mono runtime.

- Changes offsets-tool.py's user interface slightly; '--include-prefix' is
renamed to '--prefix' and may be specified multiple times. While this latter
feature isn't necessary to build a Mono cross compiler today, because we don't
use distribution-supplied cross-compilation headers on CI, it does make it
easier to experiment with the offsets tool using arbitrary header layouts.  For
example, Debian's arm64 cross-compilation packages scatter useful header files
across `/usr/lib/gcc-cross/aarch64-linux-gnu` and `/usr/aarch64-linux-gnu`.

- Updates the docker image used for arm64 cross-compilation to a newer revision
that includes libclang (see also
https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/375).

- Adds some tests that currently fail to compile with Mono LLVM AOT to
issues.targets.

- Adds a Linux_arm64 LLVM AOT job to CI. Nothing particularly fancy is done to
build the Mono LLVM AOT cross compiler; it is built as a step in the same job
that also sends Linux_arm64 tests to Helix.

### CODE REVIEW

**Comment by naricc** [comment]
@tommcdon Do we have the arm64 capacity to actually turn this on? This is the AOT runs on arm64.

**Comment by imhameed** [comment]
I reproduced the current batch of LLVM AOT compilation failures on another arm64 machine. There are 964 failures. They all look like this:
```
MONO_ENV_OPTIONS="--aot=llvm,llvm-path=/home/imran/ms/runtime/prj/artifacts/bin/mono/Linux.arm64.Release" /home/imran/ms/runtime/prj/artifacts/tests/coreclr/Linux.arm64.Release/Tests/Core_Root/corerun /home/imran/ms/runtime/prj/artifacts/tests/coreclr/Linux.arm64.Release/Loader/classloader/generics/Instantiation/Positive/NestedInterface02/NestedInterface02.dll
Mono Ahead of Time compiler - compiling assembly /home/imran/ms/runtime/prj/artifacts/tests/coreclr/Linux.arm64.Release/Loader/classloader/generics/Instantiation/Positive/NestedInterface02/NestedInterface02.dll
AOTID A69B4D55-7AC9-384F-7D20-3CE99CC339F0
Executing opt: "/home/imran/ms/runtime/prj/artifacts/bin/mono/Linux.arm64.Release/opt" -f -O2 -disable-tail-calls -place-safepoints -spp-all-backedges -o "mono_aot_Im4oW8/temp.opt.bc" "mono_aot_Im4oW8/temp.bc"
Executing llc: "/home/imran/ms/runtime/prj/artifacts/bin/mono/Linux.arm64.Release/llc"  -enable-implicit-null-checks -disable-fault-maps -march=aarch64 -asm-verbose=false -disable-gnu-eh-frame -enable-mono-eh-frame -mono-eh-frame-symbol=mono_aot_NestedInterface02_eh_frame -disable-tail-calls -relocation-model=pic -filetype=obj -o "mono_aot_Im4oW8/temp-llvm.o" "mono_aot_Im4oW8/temp.opt.bc"
llc: /home/imran/ms/llvm-9.x/prj/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:1970: bool llvm::AArch64InstrInfo::getMemOperandWithOffsetWidth(const llvm::MachineInstr&, const llvm::MachineOperand*&, int64_t&, unsigned int&, const llvm::TargetRegisterInfo*) const: Assertion `LdSt.mayLoadOrStore() && "Expected a memory operation."' failed.
Stack dump:
0.      Program arguments: /home/imran/ms/runtime/prj/artifacts/bin/mono/Linux.arm64.Release/llc -enable-implicit-null-checks -disable-fault-maps -march=aarch64 -asm-verbose=false -disable-gnu-eh-frame -enable-mono-eh-frame -mono-eh-frame-symbol=mono_aot_NestedInterface02_eh_frame -disable-tail-calls -relocation-model=pic -filetype=obj -o mono_aot_Im4oW8/temp-llvm.o mono_aot_Im4oW8/temp.opt.bc
1.      Running pass 'Function Pass Manager' on module 'mono_aot_Im4oW8/temp.opt.bc'.
2.      Running pass 'Machine code sinking' on function '@GenInt_InstVerify_System_Type'
/home/imran/ms/runtime/prj/artifacts/bin/mono/Linux.arm64.Release/llc(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamE+0x28)[0xaaaac8dbbb28]
Aborted
AOT of image /home/imran/ms/runtime/prj/artifacts/tests/coreclr/Linux.arm64.Release/Loader/classloader/generics/Instantiation/Positive/NestedInterface02/NestedInterface02.dll
```
I'm debugging this now.

**Comment by imhameed** [comment]
Where this is at now: I have an LLVM Mono AOT cross compiler working locally and I've placed instructions for making one in mono.proj; the remaining work entails gluing all of this into the existing build tooling.

File: src/mono/mono/tools/offsets-tool/offsets-tool.py [hunk 1]
```diff
@@ -89,16 +89,18 @@ def require_emscipten_path (args):
...
-				if not os.path.isdir (args.include_prefix):
-					print ("provided path via --include-prefix (\"" + args.include_prefix + "\") doesn't exist.", file=sys.stderr)
-					sys.exit (1)
-				self.target_args += ["-I", args.include_prefix + "/include"]
-				self.target_args += ["-I", args.include_prefix + "/include-fixed"]
+				for prefix in args.include_prefix:
+					if not os.path.isdir (args.include_prefix):
+						print ("provided path via --include-prefix (\"" + args.include_prefix + "\") doesn't exist.", file=sys.stderr)
+						sys.exit (1)
+					self.target_args += ["-I", args.include_prefix + "/include"]
```
Comment (vargaz) [inline src/mono/mono/tools/offsets-tool/offsets-tool.py:102]: This should be 'prefix'.

File: src/mono/mono/tools/offsets-tool/offsets-tool.py [hunk 2]
```diff
@@ -89,16 +89,18 @@ def require_emscipten_path (args):
...
+					self.target_args += ["-I", args.include_prefix + "/include-fixed"]
 			else:
```
Comment (vargaz) [inline src/mono/mono/tools/offsets-tool/offsets-tool.py:104]: Ditto.

**Comment by imhameed** [comment]
This now depends on: https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/375

**Review by trylek — APPROVED** [review]
LGTM, thank you!

**Pull request merged**